### DISCUSSION TITLE
Symfony Process is unable to use WSL commands

### DISCUSSION THREAD

**Discussion by sero583** [symfony/symfony#46470] [Q&A]
### Symfony version(s) affected

any

### Description

I tried to use the [spatie/docker](https://github.com/spatie/docker) dependency, to be able to manage docker containers by PHP code. However during execution on my local Windows 10 machine I ran all the time into an error, which told me that the command was not found. I have installed WSL (Windows Subsystem Linux, using Ubuntu) and docker, my CLI actually supports it. Then I thought, it might be PHP, which is unable to process WSL commands. I tested the simple command ```docker version``` using exec() and shell_exec(), both gave me the expected output. Then I was curios about, how it get's handled by [spatie/docker](https://github.com/spatie/docker). As I was searching in their source code, I saw they use for this the Symfony class Process. So I thought, I should do the same. During research I found out, that Symfony Process uses the vanilla CLI of windows, which has no WSL activated and that's why the docker command on my local windows machine didn't work. I even tried starting during the process WSL in the CLI by running the command ```wsl``` and then executing ```docker version```. Command still wasn't found. The exit code, which I got from Process was 127, which means such as command not found. Symfony could process the ```ls``` of Windows, which is ```dir```, but couldn't process ```ls```, due to it being a linux command. 

### How to reproduce

Run Symfony Process on WSL and try to run a WSL-command, such as ```ls``` or ```docker```, when installed, there.

### Possible Solution

Use different command execution handling on windows.

**Comment by sero583** [comment]
More Infos to this issue: I discovered, that WSL commands work. However, they don't work the usual way.

If I would try to execute ```docker version```, usually I would use for with Symfony Process the following code (Variant one):
```PHP
Process::fromShellCommandline("docker version")->run();
```

However, this will not work and the process will end up telling that it didn't find the command. When running ```docker version``` as a each-part-split-up command array, it somehow works. E. g. like this (Variant two):
```PHP
Process::fromShellCommandline(["docker", "version"])->run();
```
This will work, but as far as I know, version would be the following command and for the command ```docker version``` I should use variant one, which is not working.

**Comment by sero583** [comment]
Additional information: As seeing from the source, Symfonys Process is using ```proc_open```. After researching I found out that ```proc_open``` is able to run WSL commands, so the extension is not the issue. My best guess is, that Process might use it in a way, in which it isolates itself from WSL.

**Comment by derrabus** [comment]
I'm moving this to support because I don't think that there is a bug here. The process component is not aware of WSL or any other virtualized environment. If you run your PHP application on Linux inside your WSL container, it should be able to run all command that are available there. If you run it on Windows however, it doesn't know or guess that you meant to run a WSL process.

**Comment by nicolas-grekas** [comment]
As the name of the method suggests, variant one uses a shell, aka the command is parsed by `cmd.exe`. That's why it's not compatible with WSL I guess.

Variant two (which should be `new Process(["docker", "version"])->run()` btw) skips the shell. That's why it works I guess.
